src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,255 @@
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse* response) {
+  if (request.service_name() != service_name_) {
+    return Status(Code::INVALID_ARGUMENT,
+                  (string("Invalid service name: ") + request.service_name() +
+                   string(" Expecting: ") + service_name_));
+  }
+
+  if (!request.has_allocate_operation()) {
+    return Status(Code::INVALID_ARGUMENT,
+                  "allocate operation field is required.");
+  }
+
+  if (!cache_) {
+    // By returning NO_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+  if (!lookup.Found()) {
+    // Create a temporary element which will be remained in the cache until
+    // actual response arrives
+    ::google::api::servicecontrol::v1::AllocateQuotaResponse temp_response;
+    InternalCacheResponse(request, temp_response, true);
+    GOOGLE_LOG(INFO) << "Inserted a new temporary cache for aggregation";
+
+    // By returning NOT_FOUND, caller will send request to server.
+    return Status(Code::NOT_FOUND, "");
+  }
+
+  lookup.value()->Aggregate(request);
+
+  *response = lookup.value()->quota_response();
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// Caches a response from a remote Service Controller AllocateQuota call.
+::google::protobuf::util::Status QuotaAggregatorImpl::CacheResponse(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response) {
+  if (!cache_) {
+    return ::google::protobuf::util::Status::OK;
+  }
+
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer stack_buffer(this);
+  MutexLock lock(cache_mutex_);
+  AllocateQuotaCacheRemovedItemsHandler::StackBuffer::Swapper swapper(
+      this, &stack_buffer);
+
+  InternalCacheResponse(request, response, false);
+
+  return ::google::protobuf::util::Status::OK;
+}
+
+// This method must be called behind the cache_mutex_ lock
+void QuotaAggregatorImpl::InternalCacheResponse(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
+    const ::google::api::servicecontrol::v1::AllocateQuotaResponse& response,
+    bool is_refreshing) {
+  if (!cache_) {
+    return;
+  }
+
+  string request_signature = GenerateAllocateQuotaRequestSignature(request);
+  QuotaCache::ScopedLookup lookup(cache_.get(), request_signature);
+
+  CacheElem* cache_elem = new CacheElem(response, SimpleCycleTimer::Now());
+  cache_elem->set_signature(request_signature);
+
+  if (lookup.Found()) {
+    if (lookup.value()->is_aggregated()) {
```

Comment:
New response should stay for the cache refresh duration. To update the refresh time of the element, previous one should be removed and inserted new response to the cache.